src/quantization.cpp

```diff
@@ -112,80 +86,153 @@ instruction_ref insert_quant_ins(module& modl,
+            {
+                input_fp16 = input->inputs().front();
                 converted_inputs.push_back(input_fp16);
             }
             else
             {
-                converted_inputs.push_back(input);
+                input_fp16 = insert_quant_ins(m, ins, input, shape::half_type, map_fp16);
+                converted_inputs.push_back(input_fp16);
             }
         }
 
-        // no change for the input, go to the next instruction
-        if(inputs == converted_inputs)
+        auto mod_inputs = ins->module_inputs();
+        for(auto*& smod : mod_inputs)
         {
-            continue;
+            quantize_fp16(*smod, ins_names, map_fp16, true);
         }
 
         auto op        = ins->get_operator();
-        auto ins_shape = compute_shape(op, converted_inputs);
-        if(ins_shape.type() != orig_type)
+        auto ins_shape = compute_shape(op, converted_inputs, mod_inputs);
+        if(ins_shape != orig_shape)
         {
-            // check the dead code case to avoid assert
-            bool output_empty  = ins->outputs().empty();
-            auto ins_orig_type = mm->insert_instruction(
-                std::next(ins), make_op("convert", {{"target_type", orig_type}}), ins);
-            if(!output_empty)
+            // tuple type, followed by get_tuple_elem
+            if(ins_shape.type() == shape::tuple_type)
             {
-                mm->replace_instruction(ins, ins_orig_type);
+                auto outputs = ins->outputs();
+                for(auto out : outputs)
+                {
+                    auto out1 = m.insert_instruction(
+                        std::next(out),
+                        make_op("convert", {{"target_type", out->get_shape().type()}}),
+                        out);
+                    m.replace_instruction(out, out1);
+                }
+            }
+            else
+            {
+                // check the dead code case to avoid assert
+                auto ins_orig_shape = m.insert_instruction(
+                    std::next(ins), make_op("convert", {{"target_type", orig_shape.type()}}), ins);
+                m.replace_instruction(ins, ins_orig_shape);
             }
         }
-
-        mm->replace_instruction(ins, op, converted_inputs);
+        m.replace_instruction(ins, op, converted_inputs, mod_inputs);
     }
 }
 
+// This function is to convert any instructions specified in the input
+// from double or float to float16 by inserting a convert operator.
+// For the conversion, there could be cases of overflowing, but it
+// is very rare in the area of deeping learning, so we just do a
+// truncate of the input to get the fp16.
+void quantize_fp16(program& prog, const std::vector<std::string>& ins_names)
+{
+    auto* mm = prog.get_main_module();
+    std::unordered_map<instruction_ref, instruction_ref> map_fp16;
+    quantize_fp16(*mm, ins_names, map_fp16, false);
```

Comment:
This seems backwards. Shouldn't we start at the leaf modules first and then do the main module last?

Reply:
Actually, `run_passes` already does this. Can we convert this to a pass and then pass it to `run_passes`?

```cpp
void quantize_fp16(program& prog, const std::vector<std::string>& ins_names)
{
    run_passes(prog, {quantize_fp16_pass{ins_names}});
}
```